File: bkp/HackerRank/Problem-Solving/easy/grading-students/js/grade_v1.js
/**
 * Given some integer `num`, find its next multiple of `multiplier`.
 *
 * ASSUME: Both values are integers. If input are not integers, the
 *
 * ASSUME: The multiplier is not zero, which would cause `NaN` to
 * be returned.
 *
 * The result is always an integer, unless it returns `NaN` due to
 * incorrect inputs.
 *
 * @param {number} num
 * @param {number} multiplier
 * @returns {number}
 */
function getNextMultOf(multiplier, num) {
  return Math.ceil(num / multiplier) * multiplier;
}

/**
 * Conditionally rounds `grade` according to the rules.
 *
 * - If grade < 38, do nothing.
 * - If grade >= 38, then:
 *   - If difference between next multiple of the grade and the grade
 *     is less than 3, round up to the next multiple of 5.
 *   - Otherwise, do nothing.
 *
 * @param {number} grade
 * @returns {number}
 */
function round(grade) {
  const newGrade = getNextMultOf(5, grade);

  return (newGrade - grade < 3)
    ? newGrade
    : grade;
}

/**
 * Computes the student grades according to the grading rules.
 *
 * @param {Array<number>} grades
 * @returns {Array<number>}
 */
function gradeStudents(grades) {
  const newGrades = [];
  let i = 0, grade;

  while ((grade = grades[i++]) !== undefined)
    newGrades.push(grade < 38 ? grade : round(grade));

  return newGrades;
}

export { round, gradeStudents };

File: src/hackerrank/problem-solving/easy/grading-students/c/grade_v1.c
#include <stdio.h>
#include <math.h>

#define GRADES_LEN 8
#define MAX_GRADES 100

/**
 * Finds the next multiple of `n` given the multiplier `m`.
 *
 * - T.C: O(1).
 * - S.C: O(1).
 */
short next_mult_of(short m, short n) {
  return ceil(n / (double) m) * m;
}

/**
 * Apply the grading logic to the grades.
 *
 * - T.C: O(n).
 * - S.C: O(n).
 */
short* grade(short grades_len, short* grades, short* res_len) {
  short i, grade;
  static short res[MAX_GRADES];

  for (
    i = 0, grade = *(grades + i);
    i < grades_len;
    ++i, grade = *(grades + i)
  ) {
    if (grade < 38) {
      res[i] = grade;
      continue;
    }

    short new_grade = next_mult_of(5, grade);

    res[i] = (new_grade - grade < 3) ? new_grade : grade;
  }

  *res_len = i;

  return res;
}

int main(int argc, char* argv[]) {
  short i;
  // short xs[GRADES_LEN] = { 3, 37, 38, 39, 40, 41, 89, 98 };
  short xs[MAX_GRADES] = { 4, 73, 67, 38, 33 };
  short* result;
  short result_count;

  result = grade(5, xs, &result_count);

  for (i = 0; i < result_count; ++i) {
    printf("%hd\n", *(result + i));
  }

  return 0;
}

